File: c/data_structures/tree/traversal/postorder.c
#include "../binaryTree.c"

// Perform a postorder traversal of the BST where the result is: { 25, 75, 50, 110, 125, 175, 150, 100 }
//
//
//              __ 100 __
//             /         \
//         50              150
//       /    \          /     \
//     25      75      125     175
//                   /
//                 110
//
//

int n = 0;
void postorder(int tree[], struct node *root) {
    if (!root)
        return;

    if (root->left)
        postorder(tree, root->left);

    if (root->right)
        postorder(tree, root->right);

    *(tree + n) = root->value;
    n++;

    return;
}

int main(void) {
    int tree[8];

    struct node *root = makeNode(100);
    struct node *a = makeNode(50);
    struct node *b = makeNode(150);
    struct node *c = makeNode(25);
    struct node *d = makeNode(75);
    struct node *e = makeNode(125);
    struct node *f = makeNode(175);
    struct node *g = makeNode(110);

    root->left = a;
    root->right = b;

    a->left = c;
    a->right = d;

    b->left = e;
    b->right = f;

    e->left = g;

    postorder(tree, root);

    printf("{ ");
    for (n = 0; n < 8; n++, *(tree + n)) {
        printf("%d ", tree[n]);
    }
    printf(" }");

    return 0;
}

File: python/tree/traversal/postorder.py
import ipdb

from Tree import TreeNode


# Recursive
def postorder(root, visited=[]):
    if not root:
        return

    ipdb.set_trace()
    postorder(root.left)
    postorder(root.right)
    visited.append(root.value)

    return visited


# Iterative
#def postorder(root):
#    stack = []
#    current = root
#    visited = []
#
#    while current or stack:
#        ipdb.set_trace()
#        while current:
#            stack.append(current)
#            current = current.left
#
#        current = stack.pop()
#
#        while current:
#            stack.append(current)
#            current = current.right
#
#        current = stack.pop()
#        visited.append(current.value)
#        current = current.right
#
#    return visited


# Iterative, a queue and a stack.
#def postorder(root):
#    stack = [root]
#    # `queue` is a queue.
#    queue = []
#
#    while stack:
#        node = stack.pop()
#        queue.insert(0, node.value)
#
#        if node.left:
#            stack.append(node.left)
#        if node.right:
#            stack.append(node.right)
#
#    return queue


#tree = Tree([1, 2, 3, 4, 5, null, 6])
root = TreeNode(1)
root.left = TreeNode(2)
root.left.left = TreeNode(4)
root.left.right = TreeNode(5)
root.right = TreeNode(3)
root.right.right = TreeNode(6)

print(postorder(root))

#               1
#             /   \
#            /     \
#           2       3
#          / \       \
#         /   \       \
#        4     5       6
#
# postorder [4, 5, 2, 6, 3, 1]
